app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt

```diff
@@ -0,0 +1,280 @@
+
+    override fun recieveHalfBlock(request: ChainService.PeerTrustChainBlock, responseObserver: StreamObserver<ChainService.PeerTrustChainBlock>) {
+        val validation = saveBlock(request) ?: return
+        val peer = request.peer
+        val block = request.block
+        // check if block matches up with its previous block
+        // At this point gaps cannot be tolerated. If we detect a gap we send crawl requests to fill
+        // the gap and delay the method until the gap is filled.
+        // Note that this code does not cover the scenario where we obtain this block indirectly,
+        // because the code does nothing with this block after the crawlRequest was received.
+        if (validation.getStatus() === ValidationResult.ValidationStatus.PARTIAL_PREVIOUS || validation.getStatus() === ValidationResult.ValidationStatus.PARTIAL ||
+                validation.getStatus() === ValidationResult.ValidationStatus.NO_INFO) {
+            Log.e(TAG, "Request block could not be validated sufficiently, requested crawler. " + validation.toString())
+            // send a crawl request, requesting the last 5 blocks before the received halfblock (if available) of the peer
+            sendCrawlRequest(peer, block.publicKey.toByteArray(), Math.max(TrustChainBlock.GENESIS_SEQ, block.sequenceNumber - 5))
+        } else {
+            val signBlock = signBlock(peer, block)
+            val returnTrustChainBlock = ChainService.PeerTrustChainBlock.newBuilder().setPeer(me).setBlock(signBlock).build()
+            responseObserver.onNext(returnTrustChainBlock)
+            responseObserver.onCompleted()
+        }
+    }
+
+
+    override fun getPublicKey(request: ChainService.Empty, responseObserver: StreamObserver<ChainService.Key>) {
+        responseObserver.onNext(createPrivateKey())
+        responseObserver.onCompleted()
+    }
+
+    private fun createPrivateKey() =
+            ChainService.Key.newBuilder().setPublicKey(ByteString.copyFrom(keyPair.public.encoded)).build()
+
+    override fun sendLatestBlocks(request: ChainService.CrawlResponse, responseObserver: StreamObserver<ChainService.Key>) {
+        for (trustChainBlock in request.blockList) {
+            saveBlock(request.peer, trustChainBlock)
+        }
+
+        responseObserver.onNext(createPrivateKey())
+        responseObserver.onCompleted()
+    }
+
+    fun saveBlock(request: ChainService.PeerTrustChainBlock): ValidationResult? = saveBlock(request.peer, request.block)
+
+    fun saveBlock(peer: ChainService.Peer, block: MessageProto.TrustChainBlock): ValidationResult? {
+        Log.i(TAG, "Received half block from peer with IP: " + peer.hostname + ":" + peer.port +
+                " and public key: " + Peer.bytesToHex(peer.publicKey.toByteArray()))
+
+        val validation: ValidationResult
+        try {
+            validation = TrustChainBlock.validate(block, dbHelper)
+        } catch (e: Exception) {
+            e.printStackTrace()
```

Comment:
Again, we should probably log this

Reply:
Also, can we catch more specific Exceptions here?

Reply:
it doesn't actually throw any exceptions... 
they wrapped it around it I have no fucking clue why...

Reply:
![](https://media.giphy.com/media/3o7aCPyuFTbB9uuady/giphy.gif)